shared/js/background/pixel.es6.js

```diff
@@ -0,0 +1,102 @@
+/**
+ *
+ * This is part of our tool for anonymous engagement metrics
+ * Learn more at https://duck.co/help/privacy/atb
+ *
+ */
+
+const load = require('./load.es6')
+const browserWrapper = require('./$BROWSER-wrapper.es6')
+const settings = require('./settings.es6')
+const parseUserAgentString = require('../shared-utils/parse-user-agent-string.es6')
+
+/**
+ *
+ * Fire a pixel
+ *
+ * @param {string} pixelName
+ * @param {...*} args - any number of extra data
+ *
+ */
+function fire () {
+    if (!arguments.length) return
+
+    let args = Array.prototype.slice.call(arguments)
+    const pixelName = args[0]
+
+    if (typeof pixelName !== 'string') return
+
+    const url = getURL(pixelName)
+
+    if (!url) return
+
+    args = args.slice(1)
+    args.push(getAdditionalParams())
+    const paramString = concatParams(args)
+
+    // Send the request
+    load.url(url + paramString)
+}
+
+/**
+ *
+ * Return URL for the pixel request
+ *
+ */
+function getURL (pixelName) {
+    if (!pixelName) return
+
+    const url = 'https://improving.duckduckgo.com/t/'
+    return url + pixelName
+}
+
+/**
+ *
+ * Return additional params for the pixel request
+ *
+ */
+function getAdditionalParams () {
+    const browserInfo = parseUserAgentString()
+    const browser = browserInfo.browser
+    const extensionVersion = browserWrapper.getExtensionVersion()
+    const atb = settings.getSetting('atb')
+    const result = {}
+
+    if (browser) result.browser = browser
+    if (extensionVersion) result.extensionVersion = extensionVersion
+    if (atb) result.atb = atb
+
+    return result
+}
+
+/**
+ *
+ * @param {array} args - data we need to append
+ *
+ */
+function concatParams (args) {
+    args = args || []
+
+    let paramString = ''
+
+    args.forEach((arg) => {
+        // append keys if object
+        if (typeof arg === 'object') {
+            paramString += Object.keys(arg).reduce((params, key) => {
+                const val = arg[key]
+                if (val) return params + '&' + key + '=' + val
+            }, '')
+        } else if (arg) {
+            // otherwise just add args separated by _
+            paramString += '_' + arg
+        }
```

Comment:
There's a couple of things I spotted here:

1. This produces malformed URLs, e.g. `/t/ep&foo=bar&bar=baz`. The querystring param needs to be separated with a `?`, so something like: `/t/ep?foo=bar&bar=baz`.

2. This is arguably an edge case, but if you pass an object _before_ a string argument, like this:

```js
pixel.fire('ep', { foo: 'bar' }, 'baz')
```

you'll end up with something like:

```
/t/ep_&foo=bar_baz
```

whereas our internal pixel class would produce something like:

```
/t/ep_baz?foo=bar
```

Reply:
Also, I really don't trust browser caching, can we add a `Math.random()` same way we do in ATB?

https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/fa8b37dde86812c7343e92256a961b732ff8f129/shared/js/background/atb.es6.js#L37-L38